[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25896 [pytorch] update build_android.sh to not build host protoc for libtorch
* #25894 [pytorch] remove pthreadpool dependency in aten/CMake
* #25817 [pytorch][mobile] disable backward function codegen if disable_autograd is set
* #25816 [pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute
* **#25697 [pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile**
* #25815 [pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH

Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp,
  VariableTypeManual.cpp from the build process. Still keep variable_factories.h
  as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp)
  with C10_MOBILE (technically we should use a dedicated c macro but its
  maintenance cost is higher than cmake macro as we have several build systems
  to change).
* Pass --disable-autograd flag to codegen script, which will stop generating
  Functions/VariableType code. And for variable_factories.h it will stop
  generating tracing code.

Edit: in this diff we will keep Functions.h/cpp to avoid changing source code.

Why we need this change if it's already not calling VariableType and autograd
stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's
relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar
functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25697

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 committed:
[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile

Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

Test Plan:
- will check CI;
- test mobile build in sample app;

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

Test Plan:
- will check CI;
- test mobile build in sample app;

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp, VariableTypeManual.cpp from the build process. Still keep variable_factories.h as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp) with C10_MOBILE (technically we should use a dedicated c macro but its maintenance cost is higher than cmake macro as we have several build systems to change).
* Pass --disable-autograd flag to codegen script, which will stop generating Functions/VariableType code. And for variable_factories.h it will stop generating tracing code.

Why we need this change if it's already not calling VariableType and autograd stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

gchanan commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -139,6 +142,7 @@ const std::shared_ptr<Node>& Variable::grad_fn() const {
     if (!diff_view_meta->grad_fn_ && !diff_view_meta->base_.requires_grad()) {
       return diff_view_meta->grad_fn_;
     }
+#ifndef C10_MOBILE
```
as mentioned over chat, this part seems difficult to maintain.

The rest doesn't seem to be adding that much complexity, because we already expose AutoNonVariableType as part of the pseudo-public interface and you are just slicing along that border.

ezyang replied:
The reason this is commented out is, I assume, because of the direct reference to `generated::AsStridedBackward`? That doesn't seem too bad to me.

ljk53 replied:
Yes, generated::AsStridedBackward is the main blocker.
Ok, I updated the PR to keep the Functions.h/cpp for now to keep generated::AsStridedBackward and remove #if/#endif here.
In next PR I can update the codegen script to only generate this single function for mobile build - Does it sound better than the gating? @gchanan

ezyang commented:
> It's trying to reduce static library size for iOS build, for which it's relatively harder to strip size with linker approach.

I wanted to claim that it should (morally) be possible to reduce the size of a static library by specifying what symbols you want to keep, and then doing a DCE with those symbols as roots. But I spent a while googling and it doesn't seem like this is a thing people actually do.

I want to confirm that the static library size for iOS is a cosmetic issue: in the end, you'll still end up with an executable of the same size--it's just that it looks bad if your static library looks big, is that right?

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp, VariableTypeManual.cpp from the build process. Still keep variable_factories.h as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp) with C10_MOBILE (technically we should use a dedicated c macro but its maintenance cost is higher than cmake macro as we have several build systems to change).
* Pass --disable-autograd flag to codegen script, which will stop generating Functions/VariableType code. And for variable_factories.h it will stop generating tracing code.

Why we need this change if it's already not calling VariableType and autograd stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 commented:
> I want to confirm that the static library size for iOS is a cosmetic issue: in the end, you'll still end up with an executable of the same size--it's just that it looks bad if your static library looks big, is that right?

Correct

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp, VariableTypeManual.cpp from the build process. Still keep variable_factories.h as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp) with C10_MOBILE (technically we should use a dedicated c macro but its maintenance cost is higher than cmake macro as we have several build systems to change).
* Pass --disable-autograd flag to codegen script, which will stop generating Functions/VariableType code. And for variable_factories.h it will stop generating tracing code.

Why we need this change if it's already not calling VariableType and autograd stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp,
  VariableTypeManual.cpp from the build process. Still keep variable_factories.h
  as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp)
  with C10_MOBILE (technically we should use a dedicated c macro but its
  maintenance cost is higher than cmake macro as we have several build systems
  to change).
* Pass --disable-autograd flag to codegen script, which will stop generating
  Functions/VariableType code. And for variable_factories.h it will stop
  generating tracing code.

Edit: in this diff we will keep Functions.h/cpp to avoid changing source code.

Why we need this change if it's already not calling VariableType and autograd
stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's
relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar
functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25697

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp,
  VariableTypeManual.cpp from the build process. Still keep variable_factories.h
  as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp)
  with C10_MOBILE (technically we should use a dedicated c macro but its
  maintenance cost is higher than cmake macro as we have several build systems
  to change).
* Pass --disable-autograd flag to codegen script, which will stop generating
  Functions/VariableType code. And for variable_factories.h it will stop
  generating tracing code.

Edit: in this diff we will keep Functions.h/cpp to avoid changing source code.

Why we need this change if it's already not calling VariableType and autograd
stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's
relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar
functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25697

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp,
  VariableTypeManual.cpp from the build process. Still keep variable_factories.h
  as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp)
  with C10_MOBILE (technically we should use a dedicated c macro but its
  maintenance cost is higher than cmake macro as we have several build systems
  to change).
* Pass --disable-autograd flag to codegen script, which will stop generating
  Functions/VariableType code. And for variable_factories.h it will stop
  generating tracing code.

Edit: in this diff we will keep Functions.h/cpp to avoid changing source code.

Why we need this change if it's already not calling VariableType and autograd
stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's
relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar
functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25697

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ezyang commented on tools/autograd/gen_autograd.py:
```diff
@@ -191,21 +191,19 @@ def gen_autograd(aten_path, out, autograd_dir):
 
-    # Load deprecated signatures
-    deprecated = load_deprecated_signatures(
-        aten_decls, os.path.join(autograd_dir, 'deprecated.yaml'))
```
Oh? Is this dead? Joyous day!

ljk53 replied:
It's still used by gen_autograd_python code path :)
load_deprecated_signatures() doesn't seem to have side-effect so I assume it's safe to delete since the result is not used here.

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp,
  VariableTypeManual.cpp from the build process. Still keep variable_factories.h
  as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp)
  with C10_MOBILE (technically we should use a dedicated c macro but its
  maintenance cost is higher than cmake macro as we have several build systems
  to change).
* Pass --disable-autograd flag to codegen script, which will stop generating
  Functions/VariableType code. And for variable_factories.h it will stop
  generating tracing code.

Edit: in this diff we will keep Functions.h/cpp to avoid changing source code.

Why we need this change if it's already not calling VariableType and autograd
stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's
relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar
functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25697

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp,
  VariableTypeManual.cpp from the build process. Still keep variable_factories.h
  as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp)
  with C10_MOBILE (technically we should use a dedicated c macro but its
  maintenance cost is higher than cmake macro as we have several build systems
  to change).
* Pass --disable-autograd flag to codegen script, which will stop generating
  Functions/VariableType code. And for variable_factories.h it will stop
  generating tracing code.

Edit: in this diff we will keep Functions.h/cpp to avoid changing source code.

Why we need this change if it's already not calling VariableType and autograd
stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's
relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar
functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25697

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)

ljk53 committed:
Update on "[pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile"


Summary:
This is the first of a series of changes to reduce build size by cutting
autograd functions from mobile build.

When INTERN_DISABLE_AUTOGRAD is set:
* On CMake side we exclude Functions.h/cpp, VariableType*.h/cpp,
  VariableTypeManual.cpp from the build process. Still keep variable_factories.h
  as we rely on it to create variables instead of tensors.
* In source code we gate a couple autograd references (in autograd/variable.cpp)
  with C10_MOBILE (technically we should use a dedicated c macro but its
  maintenance cost is higher than cmake macro as we have several build systems
  to change).
* Pass --disable-autograd flag to codegen script, which will stop generating
  Functions/VariableType code. And for variable_factories.h it will stop
  generating tracing code.

Edit: in this diff we will keep Functions.h/cpp to avoid changing source code.

Why we need this change if it's already not calling VariableType and autograd
stuff with USE_STATIC_DISPATCH=ON for mobile?
It's trying to reduce static library size for iOS build, for which it's
relatively harder to strip size with linker approach.

Why we need make involved change into codegen script?
There isn't a global config system in codegen - autograd/env.py provides similar
functionality but it says not adding anything there.

Test Plan:
- will check CI;
- test mobile build in sample app;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25697

Differential Revision: [D17202733](https://our.internmc.facebook.com/intern/diff/D17202733)
